File: C/MergeKSortedLists.c
// https://leetcode.com/problems/merge-k-sorted-lists/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
void mergeKNodes(struct ListNode* node, struct ListNode** lists, int listsSize) {    
    bool allNull = true;
    
    for (int i=0; i<listsSize; i++) {
        if (lists[i] != NULL) {
            allNull = false;
            break;
        }
    }
    
    if (allNull) {
        return;
    }
    //find min value
    int index = 0;
    int min = INT_MAX;
    
    for (int i=0; i<listsSize; i++) {
        if((lists[i] != NULL) && (lists[i]->val <= min)) {
            index = i;
            min = lists[i]->val;
        }
    }
    //add it to list and advance
    node->next = lists[index];
    lists[index] = lists[index]->next;
        
    //call recursive method
    mergeKNodes(node->next, lists, listsSize);
}


struct ListNode* mergeKLists(struct ListNode** lists, int listsSize){
    struct ListNode* head = malloc(sizeof(struct ListNode));
    head->next = NULL;
    mergeKNodes(head, lists, listsSize);
    return head->next;
}

File: Java/MergeKSortedLists.java
/**
 * https://leetcode.com/problems/merge-k-sorted-lists/
 */

 /**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode head = new ListNode();
        mergeKNodes(head, lists);
        return head.next;
    }
    
    private void mergeKNodes(ListNode node, ListNode[] lists) {
        boolean allNull = true;
        for (int i=0; i<lists.length; i++) {
            if (lists[i] != null) {
                allNull = false;
                break;
            }
        }
        if (allNull) {
            return;
        }
        //find min value
        int index = 0; 
        int min = Integer.MAX_VALUE;
        for (int i=0; i<lists.length; i++) {
            if ((lists[i] != null) && (lists[i].val <= min)) {
                index = i;
                min = lists[i].val;
            }
        }
        
        //add it to list and advance
        node.next = lists[index];
        lists[index] = lists[index].next;
        
        //call recursive method
        mergeKNodes(node.next, lists);
    }
}